Modified KanesMethod to accept empty load

yidingjiang opened the pull request:
Fixes to #10841. 
I have made changes to the kanes_method in KanesMethod to accept 
a load that is either None, an empty tuple or a list containing one empty
tuple. The new fr will be calculated with a dummy tuple of the inertial
frame of the system and the zero vector in that frame. The new fr will be
a Matrix of zero column as desired.

yidingjiang committed:
mechanics: modified KanesMethod to accept empty load

I have made changes to the kanes_method in KanesMethod to accept
a load that is either None, an empty tuple or a list containing one empty
tuple. The new fr will be calculated with a dummy tuple of the inertial
frame of the system and the zero vector in that frame. The new fr will be
a Matrix of zero column as desired.

yidingjiang committed:
mechanics: modification to KanesMethod to take empty load

Updating from the previous version where the load is checked in kanes_equation.
Now the check happens within _form_fr, which makes the logic clearer.
Furthermore, assert statements have been added to the one_dof tests.

moorepants commented:
This is a hack fix. I think you want to focus on fixing the `sympy/physics/mechanics/functions.py:_f_list_parser` function.

Also see #10843

yidingjiang committed:
mechanics: updated modified KanesMethod

Improvement from previous version. Got rid of previously hacky if-checks.
Modified _f_list_parser to handle cases where the load is free. However,
the case where the forcing list contains an empty tuple is yet to be handled.
But the case is extremely rare and does seem to be taken consideration
previously. Also a small fix to the documentation of _f_list_parser.

yidingjiang commented:
@moorepants Thank you very much! I have applied changes _f_list_parser and made my commit. However, I was just wondering if the case where an empty tuple in the forcing list is a case frequent enough that needs to be handled (based on my observation, it seems extremely rare that it could happen.)

moorepants commented:
Ok, this looks pretty good. There are many cases where you have force and torque free systems. Most any satellite falls into this situation. Or anything you'd like to do in a gravity-free area.

moorepants commented on sympy/physics/mechanics/functions.py:
```diff
@@ -552,14 +552,20 @@ def _f_list_parser(fl, ref_frame):
+        for pair in fl:
+            if not pair:
+                # TODO: Handle the case where there is an empty tuple in an
+                # otherwise non-empty forcing list (which is a very rare case)
```
Oh, I see what you mean. If someone passes in something like `((), (P, force))`. I don't think that we need to support this. You can specify in the docstring of kanes_equations that the loads should be simply None. We could make it only support:

1) `None`
2) an iterable of point/frame, vector tuples

Supporting any kind of empty iterable is actually a little odd and redundant.

In the other thread I mentioned that the API should ideally be:

``def kanes_equations(bodies_and_particles, loads=None):`

But that would require a deprecation and change in the API due to the order change of the arguments. You will always have at least one particle or body but you may have no loads.

We could do this though, to stay backwards compatible:

``def KanesMethod.kanes_equations(loads=None, bodies_and_particles=None):`

This would allow the users to do any of these:

```python
kane.kanes_equations((body1, body2, particle1))
kane.kanes_equations((body1, ), (load1, load 2))
kane.kanes_equations((body1, body 2), loads=None)
kane.kanes_equations(bodies_and_particles=(body1, body2), loads=(load1, load2))
```

These would fail:

```python
kane.kanes_equations((body1, body2), loads=())
kane.kanes_equations((body1, body2), loads=[()])
```

BTW, I think we need a better name for "bodies_and_particles", something succint. But I guess a particle can be considered an infinitesimal sized body. I also prefer "loads" to forces, because the user can pass in forces or torques.

I think what I show above is the best we can do with the API and it will follow Python's Zen: "There should be one-- and preferably only one --obvious way to do it."

yidingjiang commented:
Thank you for the very detailed explanation. That makes a lot of sense. I will make the mentioned changes tonight when I get back home. If there isn't any other issues I think it will be ready for merge? :)

moorepants commented:
I don't know if it will be ready to merge until the new code is reviewed.

Please make unit tests for the API designs that I show above and if they all pass, then this is likely to be ready to be merged. You'll need to update the docstring of `kanes_equations` to reflect this.

yidingjiang commented:
@moorepants correct me if I'm wrong, but for kanes_equation((body,particle)), it seems that the tuple is assumed to be load rather than bodies with the current implementation. At least it is this way when I tested on Python 3.4.

moorepants commented:
This is the current implementation:

```def kanes_equations(self, FL, BL):```

You need to keep the order of the arguments the same so that we don't have to deprecate the functionality. But you can make both of them keyword arguments (in the same order). Do do what I have written above.

moorepants commented:
Oh, I see what you mean. You'll have to do this:

`kanes_equations(None, bodies)`

or

`kanes_equations(loads=None, bodies)`

Which is backwards compatible but a little annoying to have to type. If we make them both keyword args you can do this:

`kanes_equations(bodies=(body1, body2))`

Which is a little nicer I suppose. Either way, you can have args or kwargs.

yidingjiang committed:
mechanics: updated input constraints of kanes_equation

Modified the input format restriction of kanes_equation.
Now the method only accepts:
1. a non-empty interable of tuples of frame+torque or point+force.
2. None for no constraints.
Everything else will raise exception. A new test suites test_input_format
has been added to test these changes. The docstring of kanes_equation
has been modified accordingly as well.

yidingjiang commented:
I have pushed the changes and test. I think the input format 'kanes_equation((body1,body2), loads=None)' is actually not possible because it raises a syntax error due to the aforementioned problem.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,10 +696,12 @@ def kanes_equations(self, FL, BL):
         ==========
 
-        FL : list
+        loads : list
```
Make this:

`loads: iterable`

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,10 +696,12 @@ def kanes_equations(self, FL, BL):
-        BL : list
+            Must be either a non-empty list of tuples or None which corresponds
+            to a system with no constraints.
+        bodies : list
```
And this:

`bodies: iterable`

moorepants replied:
What happens if the default "None" is supplied? Please explain.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,10 +696,12 @@ def kanes_equations(self, FL, BL):
 
-        FL : list
+        loads : list
             Takes in a list of (Point, Vector) or (ReferenceFrame, Vector)
```
This should not reference "list". It should say something like: "An iterable of ..."

moorepants replied:
Also mention what happens if you use the default: "None"

moorepants commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -281,3 +281,51 @@ def test_parallel_axis():
+    # test for input format kane.kanes_equations(None, (body1, body2, particle1))
+    assert KM.kanes_equations(None, BL)[0] == Matrix([0])
+    # test for input format kane.kanes_equations(bodies=(body1, body 2), loads=None)
+    assert KM.kanes_equations(bodies=BL, loads=None)[0] == Matrix([0])
```
also test: `kanes_equations(bodies=BL)`

moorepants replied:
Test an input with tuples instead of lists too, to make sure it works with both iterable types.

moorepants commented:
I'm thinking about this more and I wonder if we should just fix this properly now. We could do something like:

```python

def kanes_equations(self, bodies, loads=None):
    if bodies contains points/frames and vector tuples or is None:
        # This switches the order if they use the old way.
        bodies, loads = loads, bodies
        raise DeprecationWarning("The API has changed and will be deprecated, loads is now the second argument and optional.")
    
```

This would change the API to the ideal and simpler API, but we'd have to deprecate this for the next version of SymPy.

This will be annoying to users but it isn't too hard to fix and it gives us a better API for the long run.

What do you think?

yidingjiang commented:
I think it's a very good idea because it is more intuitive and concise. To be honest, I don't think setting the initial value to None makes a difference since we still need to put None into the argument. I will fix this and other changes either when I get home tonight or tommorow night since I need to finish my GSoC application.

moorepants commented:
@sympy/mechanics Does anyone think we shouldn't change the API here to the one above?

yidingjiang committed:
mechanics: fixing minor error from previous commit

Fixing trailing spaces in test_kane.py.

yidingjiang commented:
Should I switch the arguments right now? This is not a trivial change so I would like to confirm before I make any substantial changes.

yidingjiang committed:
deprecated api

yidingjiang committed:
mechanics: updated docstring and modified test

Changed the doctring of kanes_equation for to be more general
and accurate. Modified test_input_format to test input of tuples
instead of all lists.

yidingjiang commented:
@moorepants I have not yet switched the arguments. Should I do that or is this okay for now?

moorepants commented:
Yes, let's switch the arguments, make the ideal api, and have the deprecation warning.

yidingjiang committed:
mechanics: change the kanes_equation api to be ideal

Switched the argument order of kanes_equation so that
now the body list is the first argument and loads the second.
Raise a deprecation warning and switch the argument if the first
item of body list is a tuple. I believe no additional check is needed
because the rest of the code already does the check.
Also updated all the unit test and added ignore deprecation warning
to all of them.

yidingjiang committed:
Merge branch 'master' into new_patch

yidingjiang committed:
Merge branch 'deprecate' into new_patch

yidingjiang committed:
mechanics: fixed minor errors in kanes_equations

Fix a bracket issue.

yidingjiang committed:
Merge branch 'new_patch' of https://github.com/yidingjiang/sympy into new_patch

yidingjiang committed:
Merge branch 'master' into new_patch

yidingjiang committed:
Merge branch 'master' into new_patch

yidingjiang commented:
Do Travis run tests different from the ones I have? I have modified the api and also added warning ignore to all the tests. All the test for kanes and linearize work locally, but are failing for travis. How should I fix this?

moorepants commented:
Travis runs the tests on a bunch of different versions. So yes it runs things differently than your machine. If you click details you can investigate.

You need to design your tests to handle the dep warnings (which cause travis to error), e.g.:

```python
    # The old linearizer raises a deprecation warning, so catch it here so
    # it doesn't cause py.test to fail.
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=SymPyDeprecationWarning)
        F_A_old, F_B_old, r_old = KM.linearize()
```

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,20 +696,26 @@ def kanes_equations(self, FL, BL):
-
+            Must be either a non-empty iterable of tuples or None which corresponds
+            to a system with no constraints.
+        bodies : iterable
```
Put this above loads in the docstring to match the new order of the args.

yidingjiang commented:
I already added all the `with warnings.catch_warnings(): `to all occurrences of kanes_equations in the test suites, which is why all the tests are passing on my machine locally, but I am not sure if Travis has that.

yidingjiang commented:
I already added the `with warnings.catch_warnings(): ...` to all occurrences of 'kanes_equation(...,...)` in the tests under `sympy/physics/mechanics/test`. Are there more files that should be modified?

yidingjiang commented:
I already added the with warnings.catch_warnings(): ... to all occurrences of 'kanes_equation(...,...)in the tests under `sympy/physics/mechanics/test`. Are there more files that should be modified?

yidingjiang commented:
I already added the with `warnings.catch_warnings(): ...` to all occurrences of `kanes_equation(...,...)` in the tests under `sympy/physics/mechanics/test`. Are there more files that should be modified? It seems that they are not reflected in the Travis' tests.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,20 +696,26 @@ def kanes_equations(self, FL, BL):
+            # This switches the order if they use the old way.
+            bodies, loads = loads, bodies
+            SymPyDeprecationWarning('The API has changed and will be deprecated, '
+                    'loads is now the second argument and optional.').warn()
```
Let's make this more defined:

```python
SymPyDeprecationWarning(value='The API for kanes_equations() has changed such that the loads (forces and torques) are now the second argument and is optional with None being the default.`,
                                            feature='The kanes_equation() argument order',
                                            last_supported_version="1.X.X",
                                            useinstead="switched argument order to update your code, For example: kanes_equations(loads, bodies) > kanes_equations(bodies, loads).",
                                            issue=10XXX,
                                            deprecated_since_version="1.0")
```

We need to open an issue for the deprecation of this argument change and reference in the warning.

moorepants commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -278,6 +294,60 @@ def test_parallel_axis():
+    rhs = MM.inv() * forcing
+    assert expand(rhs[0]) == expand((-k1 * q1 - c1 * u1 + k2 * q2 + c2 * u2)/m)
+    assert expand(rhs[1]) == expand((k1 * q1 + c1 * u1 - 2 * k2 * q2 - 2 *
+                                    c2 * u2) / m)
```
Please test the `loads=None` somewhere here.

moorepants replied:
Is this second part necessary? Is it a new test for this issue? What does it test?

yidingjiang replied:
This just tests the most common case where the user puts in a non-empty body list and a non-empty load list.

moorepants commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -278,6 +294,60 @@ def test_parallel_axis():
+    # test for input format kane.kanes_equations(bodies=(body1, body 2), loads=None)
+    assert KM.kanes_equations(BL, loads=None)[0] == Matrix([0])
+    # test for input format kane.kanes_equations(bodies=(body1, body 2))
+    assert KM.kanes_equations(BL)[0] == Matrix([0])
```
This is good.

moorepants replied:
Does this also support an empty tuple for loads? You seem to have a function above that indicates that it would.

yidingjiang replied:
I believe I removed the support since you said there should only be one way to do stuff.

moorepants replied:
Ok

moorepants commented:
Yes, these tests are failing on the doctests:

`bin/doctest`

moorepants commented on sympy/physics/mechanics/functions.py:
```diff
@@ -552,14 +552,20 @@ def _f_list_parser(fl, ref_frame):
-        for obj, force in fl:
+        for pair in fl:
+            if not pair:
+                # TODO: Handle the case where there is an empty tuple in an
```
Can you make a test for these changes?

yidingjiang replied:
I think this should just error out? Since we should not support load format like `(load1, (), load2)`.

yidingjiang commented:
Doctest appears to be an executable, what exactly does it do? Can I make any changes to it?

moorepants commented:
That executable runs the doc tests. You need to change all uses of kanes_equations in the doc strings and doc tests to the new API.

yidingjiang committed:
mechanics: updated doc string and more detailed warning

Updated doc strings and warning now shows more details.

yidingjiang committed:
mechanics: updated rst tests

Change argument order inside rst tests under mechanics where
kanes_equation is used.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -267,9 +267,8 @@ def _initialize_kindiffeq_matrices(self, kdeqs):
-        if not iterable(fl):
-            raise TypeError('Force pairs must be supplied in an iterable.')
+        if fl != None and (len(fl) == 0 or not iterable(fl)):
+            raise TypeError('Force pairs must be supplied in an non-empty iterable or None.')
```
You can test for this condition with something like:

```python
from sympy.utilities.pytest import raises
raises(lambda: _form_fr('bad input')
```

Or whatever you are trying to check for. I think this would be a ValueError more than a type error, no? I guess it is both.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,20 +696,32 @@ def kanes_equations(self, FL, BL):
+                    'that the loads (forces and torques) are now the second argument '
+                    'and is optional with None being the default.',
+                    feature='The kanes_equation() argument order',
+                    last_supported_version="1.0",
```
I think this needs to be either `1.X.X` or somehow should say "the last version before 2.0". I have a question on the mailing list about this detail.

moorepants replied:
Let's just not set this kwarg. Leave it as None.

yidingjiang replied:
Correct me if I am wrong. Do you leave `last_supported_version` as None?

moorepants replied:
I think so. It is impossible to predict what the last supported version will be a priori.

yidingjiang committed:
mechanics: changed error type for _form_fr

Changed the error type raised for _form_fr on a bad
input from TypeError to ValueError and also added
corresponding test for this change.

yidingjiang commented:
I have made the ValueError change. I read the discussion on google group, it seems that we have not reached a definite answer. So in the meantime, how should I deal with the version problem.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -697,20 +697,32 @@ def kanes_equations(self, FL, BL):
+                    last_supported_version="1.0",
+                    useinstead='switched argument order to update your code, For example: '
+                    'kanes_equations(loads, bodies) > kanes_equations(bodies, loads).',
+                    issue=10945, deprecated_since_version="1.0").warn()
```
This should be deprecated since 1.1, as that will be the first version that this deprecation shows up.

yidingjiang replied:
So `last_supported_version` would be 1.0 and deprecated since 1.1?

moorepants replied:
It is deprecated since 1.1, which will be the most likely next release. It will be supported until 1.2 or 2.0 depending on the result of the mailing list conversation.

moorepants commented:
Ok, fix the last two comment about the versions in the dep warning and this can go in.

yidingjiang committed:
mechanics: corrected deprecation warning

Changed to deprecated since 1.1 and remove last supported
version.

moorepants commented:
@yidingjiang You'll need to update your branch with master to get this merged. See:

https://github.com/sympy/sympy/wiki/Development-workflow#synchronization-with-master-sympy-sympy

yidingjiang committed:
Merge branch 'master' into new_patch

yidingjiang commented:
@moorepants should I rebase?

moorepants commented:
No, just pull down master from upstream and merge it into your branch.

yidingjiang commented:
I mean as in clean up the commit history.

moorepants commented:
We typically avoid rebasing once code is public. We can merge as is.

yidingjiang commented:
I see. I guess that's fine then.

moorepants commented:
If you want to squash everything and know how to do it. You can.

moorepants commented:
All tests have passed. Merging. Thanks for this contribution!

yidingjiang commented:
Thank you very much! I learnt a lot.
